Error connecting to Confluent Kafka when running as native image

I am trying to post a message to a Kafka topic on Confluent Cloud using micronaut-kafka. When I don't build my application as a native image, everything works fine. When I do build a native image (using gradle dockerBuildNative) the application runs into an error while trying to authenticate at Confluent Cloud. I'm using plain SASL/JAAS as an authentication mechanism.

While googling the error I came across an issue in the quarkus github that looks very similar: https://github.com/quarkusio/quarkus/issues/5254. Perhaps looking into this issue will help in resolving the issue for Micronaut as well.

I tried to resolve the issue by adding a TypeHint annotation. This seemed to resolve the original error but I ran into a new one and got stuck. I will share both stacktraces. 

### Steps to Reproduce:

1. Register a free trial with Confluent Cloud. Generate a key / secret for plain SASL authentication, create a topic.
2. Use the following kafka configuration in application.yml:
```
kafka:
  bootstrap.servers: [confluent-bootstrap-server]
  security.protocol: SASL_SSL
  sasl.jaas.config: org.apache.kafka.common.security.plain.PlainLoginModule required username="[my-key]" password="[my-password]";
  sasl.mechanism: PLAIN
```
3. make a kafka producer as you would normally with micronaut-kafka. 
4. build a native image of the application. (I used gradle dockerBuildNative), have it post a mesage to Kafka.

### Expected Behaviour

Message get's posted to Confluent Kafka

### Actual Behaviour
I get the following stacktrace as soon as I try to post a message to my Kafka topic: 

```
11:14:26.910 [nioEventLoopGroup-1-3] ERROR i.m.h.s.netty.RoutingInBoundHandler - Unexpected error occurred: Error instantiating bean of type [org.apache.kafka.clients.producer.KafkaProducer]: Failed to construct kafka producer
io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type [org.apache.kafka.clients.producer.KafkaProducer]: Failed to construct kafka producer
        at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1918)
        at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:903)
        at io.micronaut.context.DefaultBeanContext.createBean(DefaultBeanContext.java:830)
        at io.micronaut.context.BeanContext.createBean(BeanContext.java:130)
        at io.micronaut.configuration.kafka.intercept.KafkaClientIntroductionAdvice.lambda$getProducer$13(KafkaClientIntroductionAdvice.java:663)
        at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
        at io.micronaut.configuration.kafka.intercept.KafkaClientIntroductionAdvice.getProducer(KafkaClientIntroductionAdvice.java:585)
        at io.micronaut.configuration.kafka.intercept.KafkaClientIntroductionAdvice.intercept(KafkaClientIntroductionAdvice.java:227)
        at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:69)
        at io.micronaut.retry.intercept.RecoveryInterceptor.intercept(RecoveryInterceptor.java:80)
        at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:69)
        at nl.amis.kafkaclient.TransactionProducer$Intercepted.sendTransaction(Unknown Source)
        at nl.amis.kafkaclient.TransactionProducer.sendTransaction(TransactionProducer.java:11)
        at nl.amis.controller.TransactionController.postMethod(TransactionController.java:25)
        at nl.amis.controller.$TransactionControllerDefinition$Intercepted.$$access0(Unknown Source)
        at nl.amis.controller.$TransactionControllerDefinition$Intercepted$$proxy0.invokeInternal(Unknown Source)
        at io.micronaut.context.AbstractExecutableMethod.invoke(AbstractExecutableMethod.java:146)
        at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:60)
        at io.micronaut.validation.ValidatingInterceptor.intercept(ValidatingInterceptor.java:106)
        at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:69)
        at nl.amis.controller.$TransactionControllerDefinition$Intercepted.postMethod(Unknown Source)
        at nl.amis.controller.$$TransactionControllerDefinition$InterceptedDefinition$$exec1.invokeInternal(Unknown Source)
        at io.micronaut.context.AbstractExecutableMethod.invoke(AbstractExecutableMethod.java:146)
        at io.micronaut.context.DefaultBeanContext$4.invoke(DefaultBeanContext.java:469)
        at io.micronaut.web.router.AbstractRouteMatch.execute(AbstractRouteMatch.java:312)
        at io.micronaut.web.router.RouteMatch.execute(RouteMatch.java:118)
        at io.micronaut.http.server.netty.RoutingInBoundHandler.lambda$buildResultEmitter$9(RoutingInBoundHandler.java:1352)
        at io.reactivex.internal.operators.flowable.FlowableDefer.subscribeActual(FlowableDefer.java:35)
        at io.reactivex.Flowable.subscribe(Flowable.java:14918)
        at io.reactivex.Flowable.subscribe(Flowable.java:14865)
        at io.micronaut.http.server.context.ServerRequestContextFilter.lambda$doFilter$0(ServerRequestContextFilter.java:62)
        at io.reactivex.internal.operators.flowable.FlowableFromPublisher.subscribeActual(FlowableFromPublisher.java:29)
        at io.reactivex.Flowable.subscribe(Flowable.java:14918)
        at io.reactivex.Flowable.subscribe(Flowable.java:14868)
        at io.micronaut.http.server.netty.RoutingInBoundHandler.lambda$buildExecutableRoute$5(RoutingInBoundHandler.java:1056)
        at io.micronaut.web.router.DefaultUriRouteMatch$1.execute(DefaultUriRouteMatch.java:80)
        at io.micronaut.web.router.RouteMatch.execute(RouteMatch.java:118)
        at io.micronaut.http.server.netty.RoutingInBoundHandler$2.doOnComplete(RoutingInBoundHandler.java:987)
        at io.micronaut.core.async.subscriber.CompletionAwareSubscriber.onComplete(CompletionAwareSubscriber.java:71)
        at io.micronaut.http.server.netty.jackson.JsonContentProcessor$1.doOnComplete(JsonContentProcessor.java:140)
        at io.micronaut.core.async.subscriber.CompletionAwareSubscriber.onComplete(CompletionAwareSubscriber.java:71)
        at java.util.Optional.ifPresent(Optional.java:183)
        at io.micronaut.core.async.processor.SingleThreadedBufferingProcessor.doOnComplete(SingleThreadedBufferingProcessor.java:48)
        at io.micronaut.jackson.parser.JacksonProcessor.doOnComplete(JacksonProcessor.java:139)
        at io.micronaut.core.async.subscriber.SingleThreadedBufferingSubscriber.onComplete(SingleThreadedBufferingSubscriber.java:70)
        at io.micronaut.http.server.netty.jackson.JsonContentProcessor.doOnComplete(JsonContentProcessor.java:165)
        at io.micronaut.core.async.subscriber.CompletionAwareSubscriber.onComplete(CompletionAwareSubscriber.java:71)
        at io.micronaut.http.netty.reactive.HandlerPublisher.complete(HandlerPublisher.java:416)
        at io.micronaut.http.netty.reactive.HandlerPublisher.handlerRemoved(HandlerPublisher.java:403)
        at io.netty.channel.AbstractChannelHandlerContext.callHandlerRemoved(AbstractChannelHandlerContext.java:979)
        at io.netty.channel.DefaultChannelPipeline.callHandlerRemoved0(DefaultChannelPipeline.java:637)
        at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:477)
        at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:423)
        at io.micronaut.http.netty.stream.HttpStreamsHandler.removeHandlerIfActive(HttpStreamsHandler.java:388)
        at io.micronaut.http.netty.stream.HttpStreamsHandler.handleReadHttpContent(HttpStreamsHandler.java:253)
        at io.micronaut.http.netty.stream.HttpStreamsHandler.channelRead(HttpStreamsHandler.java:228)
        at io.micronaut.http.netty.stream.HttpStreamsServerHandler.channelRead(HttpStreamsServerHandler.java:121)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
        at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
        at io.netty.handler.codec.http.HttpServerKeepAliveHandler.channelRead(HttpServerKeepAliveHandler.java:64)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.flow.FlowControlHandler.dequeue(FlowControlHandler.java:191)
        at io.netty.handler.flow.FlowControlHandler.read(FlowControlHandler.java:130)
        at io.netty.channel.AbstractChannelHandlerContext.invokeRead(AbstractChannelHandlerContext.java:686)
        at io.netty.channel.AbstractChannelHandlerContext.read(AbstractChannelHandlerContext.java:671)
        at io.netty.channel.ChannelDuplexHandler.read(ChannelDuplexHandler.java:103)
        at io.netty.channel.AbstractChannelHandlerContext.invokeRead(AbstractChannelHandlerContext.java:686)
        at io.netty.channel.AbstractChannelHandlerContext.read(AbstractChannelHandlerContext.java:671)
        at io.netty.channel.ChannelDuplexHandler.read(ChannelDuplexHandler.java:103)
        at io.netty.channel.AbstractChannelHandlerContext.invokeRead(AbstractChannelHandlerContext.java:686)
        at io.netty.channel.AbstractChannelHandlerContext.read(AbstractChannelHandlerContext.java:671)
        at io.netty.channel.ChannelDuplexHandler.read(ChannelDuplexHandler.java:103)
        at io.netty.channel.AbstractChannelHandlerContext.invokeRead(AbstractChannelHandlerContext.java:686)
        at io.netty.channel.AbstractChannelHandlerContext.read(AbstractChannelHandlerContext.java:671)
        at io.micronaut.http.netty.reactive.HandlerPublisher.requestDemand(HandlerPublisher.java:163)
        at io.micronaut.http.netty.stream.HttpStreamsHandler$1.requestDemand(HttpStreamsHandler.java:220)
        at io.micronaut.http.netty.reactive.HandlerPublisher$ChannelSubscription.receivedDemand(HandlerPublisher.java:547)
        at io.micronaut.http.netty.reactive.HandlerPublisher$ChannelSubscription.lambda$request$0(HandlerPublisher.java:474)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:834)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:517)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
Caused by: org.apache.kafka.common.KafkaException: Failed to construct kafka producer
        at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:434)
        at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:313)
        at io.micronaut.configuration.kafka.KafkaProducerFactory.createProducer(KafkaProducerFactory.java:59)
        at io.micronaut.configuration.kafka.$KafkaProducerFactory$CreateProducer0Definition.doBuild(Unknown Source)
        at io.micronaut.context.AbstractParametrizedBeanDefinition.build(AbstractParametrizedBeanDefinition.java:118)
        at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1876)
        ... 99 common frames omitted
Caused by: org.apache.kafka.common.KafkaException: org.apache.kafka.common.KafkaException: Could not find a public no-argument constructor for org.apache.kafka.common.security.authenticator.SaslClientCallbackHandler
        at org.apache.kafka.common.network.SaslChannelBuilder.configure(SaslChannelBuilder.java:172)
        at org.apache.kafka.common.network.ChannelBuilders.create(ChannelBuilders.java:157)
        at org.apache.kafka.common.network.ChannelBuilders.clientChannelBuilder(ChannelBuilders.java:73)
        at org.apache.kafka.clients.ClientUtils.createChannelBuilder(ClientUtils.java:105)
        at org.apache.kafka.clients.producer.KafkaProducer.newSender(KafkaProducer.java:442)
        at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:423)
        ... 104 common frames omitted
Caused by: org.apache.kafka.common.KafkaException: Could not find a public no-argument constructor for org.apache.kafka.common.security.authenticator.SaslClientCallbackHandler
        at org.apache.kafka.common.utils.Utils.newInstance(Utils.java:310)
        at org.apache.kafka.common.network.SaslChannelBuilder.createClientCallbackHandler(SaslChannelBuilder.java:293)
        at org.apache.kafka.common.network.SaslChannelBuilder.configure(SaslChannelBuilder.java:136)
        ... 109 common frames omitted
Caused by: java.lang.NoSuchMethodException: org.apache.kafka.common.security.authenticator.SaslClientCallbackHandler.<init>()
        at java.lang.Class.getConstructor0(DynamicHub.java:3349)
        at java.lang.Class.getDeclaredConstructor(DynamicHub.java:2553)
        at org.apache.kafka.common.utils.Utils.newInstance(Utils.java:308)
        ... 111 common frames omitted
```
After this, I tried to add the following TypeHInt annotation (inspired by some comment in the quarkus issue):
`@TypeHint({org.apache.kafka.common.security.plain.PlainLoginModule.class, ProducerKafkaMetricsReporter.class, SaslClientCallbackHandler.class,SaslClientCallbackHandler.class, DefaultLogin.class, AbstractLogin.DefaultLoginCallbackHandler.class})
`
With this typehint, I get the following stacktrace instead:

```
10:56:31.501 [nioEventLoopGroup-1-3] ERROR i.m.r.intercept.RecoveryInterceptor - Type [nl.amis.kafkaclient.TransactionProducer$Intercepted] executed with error: Error instantiating bean of type [org.apache.kafka.clients.producer.KafkaProducer]: Failed to construct kafka producer
io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type [org.apache.kafka.clients.producer.KafkaProducer]: Failed to construct kafka producer
        at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1918)
        at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:903)
        at io.micronaut.context.DefaultBeanContext.createBean(DefaultBeanContext.java:830)
        at io.micronaut.context.BeanContext.createBean(BeanContext.java:130)
        at io.micronaut.configuration.kafka.intercept.KafkaClientIntroductionAdvice.lambda$getProducer$13(KafkaClientIntroductionAdvice.java:663)
        at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
        at io.micronaut.configuration.kafka.intercept.KafkaClientIntroductionAdvice.getProducer(KafkaClientIntroductionAdvice.java:585)
        at io.micronaut.configuration.kafka.intercept.KafkaClientIntroductionAdvice.intercept(KafkaClientIntroductionAdvice.java:227)
        at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:69)
        at io.micronaut.retry.intercept.RecoveryInterceptor.intercept(RecoveryInterceptor.java:80)
        at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:69)
        at nl.amis.kafkaclient.TransactionProducer$Intercepted.sendTransaction(Unknown Source)
        at nl.amis.kafkaclient.TransactionProducer.sendTransaction(TransactionProducer.java:11)
        at nl.amis.controller.TransactionController.postMethod(TransactionController.java:25)
        at nl.amis.controller.$TransactionControllerDefinition$Intercepted.$$access0(Unknown Source)
        at nl.amis.controller.$TransactionControllerDefinition$Intercepted$$proxy0.invokeInternal(Unknown Source)
        at io.micronaut.context.AbstractExecutableMethod.invoke(AbstractExecutableMethod.java:146)
        at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:60)
        at io.micronaut.validation.ValidatingInterceptor.intercept(ValidatingInterceptor.java:106)
        at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:69)
        at nl.amis.controller.$TransactionControllerDefinition$Intercepted.postMethod(Unknown Source)
        at nl.amis.controller.$$TransactionControllerDefinition$InterceptedDefinition$$exec1.invokeInternal(Unknown Source)
        at io.micronaut.context.AbstractExecutableMethod.invoke(AbstractExecutableMethod.java:146)
        at io.micronaut.context.DefaultBeanContext$4.invoke(DefaultBeanContext.java:469)
        at io.micronaut.web.router.AbstractRouteMatch.execute(AbstractRouteMatch.java:312)
        at io.micronaut.web.router.RouteMatch.execute(RouteMatch.java:118)
        at io.micronaut.http.server.netty.RoutingInBoundHandler.lambda$buildResultEmitter$9(RoutingInBoundHandler.java:1352)
        at io.reactivex.internal.operators.flowable.FlowableDefer.subscribeActual(FlowableDefer.java:35)
        at io.reactivex.Flowable.subscribe(Flowable.java:14918)
        at io.reactivex.Flowable.subscribe(Flowable.java:14865)
        at io.micronaut.http.server.context.ServerRequestContextFilter.lambda$doFilter$0(ServerRequestContextFilter.java:62)
        at io.reactivex.internal.operators.flowable.FlowableFromPublisher.subscribeActual(FlowableFromPublisher.java:29)
        at io.reactivex.Flowable.subscribe(Flowable.java:14918)
        at io.reactivex.Flowable.subscribe(Flowable.java:14868)
        at io.micronaut.http.server.netty.RoutingInBoundHandler.lambda$buildExecutableRoute$5(RoutingInBoundHandler.java:1056)
        at io.micronaut.web.router.DefaultUriRouteMatch$1.execute(DefaultUriRouteMatch.java:80)
        at io.micronaut.web.router.RouteMatch.execute(RouteMatch.java:118)
        at io.micronaut.http.server.netty.RoutingInBoundHandler$2.doOnComplete(RoutingInBoundHandler.java:987)
        at io.micronaut.core.async.subscriber.CompletionAwareSubscriber.onComplete(CompletionAwareSubscriber.java:71)
        at io.micronaut.http.server.netty.jackson.JsonContentProcessor$1.doOnComplete(JsonContentProcessor.java:140)
        at io.micronaut.core.async.subscriber.CompletionAwareSubscriber.onComplete(CompletionAwareSubscriber.java:71)
        at java.util.Optional.ifPresent(Optional.java:183)
        at io.micronaut.core.async.processor.SingleThreadedBufferingProcessor.doOnComplete(SingleThreadedBufferingProcessor.java:48)
        at io.micronaut.jackson.parser.JacksonProcessor.doOnComplete(JacksonProcessor.java:139)
        at io.micronaut.core.async.subscriber.SingleThreadedBufferingSubscriber.onComplete(SingleThreadedBufferingSubscriber.java:70)
        at io.micronaut.http.server.netty.jackson.JsonContentProcessor.doOnComplete(JsonContentProcessor.java:165)
        at io.micronaut.core.async.subscriber.CompletionAwareSubscriber.onComplete(CompletionAwareSubscriber.java:71)
        at io.micronaut.http.netty.reactive.HandlerPublisher.complete(HandlerPublisher.java:416)
        at io.micronaut.http.netty.reactive.HandlerPublisher.handlerRemoved(HandlerPublisher.java:403)
        at io.netty.channel.AbstractChannelHandlerContext.callHandlerRemoved(AbstractChannelHandlerContext.java:979)
        at io.netty.channel.DefaultChannelPipeline.callHandlerRemoved0(DefaultChannelPipeline.java:637)
        at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:477)
        at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:423)
        at io.micronaut.http.netty.stream.HttpStreamsHandler.removeHandlerIfActive(HttpStreamsHandler.java:388)
        at io.micronaut.http.netty.stream.HttpStreamsHandler.handleReadHttpContent(HttpStreamsHandler.java:253)
        at io.micronaut.http.netty.stream.HttpStreamsHandler.channelRead(HttpStreamsHandler.java:228)
        at io.micronaut.http.netty.stream.HttpStreamsServerHandler.channelRead(HttpStreamsServerHandler.java:121)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
        at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
        at io.netty.handler.codec.http.HttpServerKeepAliveHandler.channelRead(HttpServerKeepAliveHandler.java:64)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.flow.FlowControlHandler.dequeue(FlowControlHandler.java:191)
        at io.netty.handler.flow.FlowControlHandler.read(FlowControlHandler.java:130)
        at io.netty.channel.AbstractChannelHandlerContext.invokeRead(AbstractChannelHandlerContext.java:686)
        at io.netty.channel.AbstractChannelHandlerContext.read(AbstractChannelHandlerContext.java:671)
        at io.netty.channel.ChannelDuplexHandler.read(ChannelDuplexHandler.java:103)
        at io.netty.channel.AbstractChannelHandlerContext.invokeRead(AbstractChannelHandlerContext.java:686)
        at io.netty.channel.AbstractChannelHandlerContext.read(AbstractChannelHandlerContext.java:671)
        at io.netty.channel.ChannelDuplexHandler.read(ChannelDuplexHandler.java:103)
        at io.netty.channel.AbstractChannelHandlerContext.invokeRead(AbstractChannelHandlerContext.java:686)
        at io.netty.channel.AbstractChannelHandlerContext.read(AbstractChannelHandlerContext.java:671)
        at io.netty.channel.ChannelDuplexHandler.read(ChannelDuplexHandler.java:103)
        at io.netty.channel.AbstractChannelHandlerContext.invokeRead(AbstractChannelHandlerContext.java:686)
        at io.netty.channel.AbstractChannelHandlerContext.read(AbstractChannelHandlerContext.java:671)
        at io.micronaut.http.netty.reactive.HandlerPublisher.requestDemand(HandlerPublisher.java:163)
        at io.micronaut.http.netty.stream.HttpStreamsHandler$1.requestDemand(HttpStreamsHandler.java:220)
        at io.micronaut.http.netty.reactive.HandlerPublisher$ChannelSubscription.receivedDemand(HandlerPublisher.java:547)
        at io.micronaut.http.netty.reactive.HandlerPublisher$ChannelSubscription.lambda$request$0(HandlerPublisher.java:474)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:834)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:517)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
Caused by: org.apache.kafka.common.KafkaException: Failed to construct kafka producer
        at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:434)
        at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:313)
        at io.micronaut.configuration.kafka.KafkaProducerFactory.createProducer(KafkaProducerFactory.java:59)
        at io.micronaut.configuration.kafka.$KafkaProducerFactory$CreateProducer0Definition.doBuild(Unknown Source)
        at io.micronaut.context.AbstractParametrizedBeanDefinition.build(AbstractParametrizedBeanDefinition.java:118)
        at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1876)
        ... 99 common frames omitted
Caused by: org.apache.kafka.common.KafkaException: java.util.MissingResourceException: Resource bundle not found sun.security.util.Resources. Register the resource bundle using the option -H:IncludeResourceBundles=sun.security.util.Resources.
        at org.apache.kafka.common.network.SaslChannelBuilder.configure(SaslChannelBuilder.java:172)
        at org.apache.kafka.common.network.ChannelBuilders.create(ChannelBuilders.java:157)
        at org.apache.kafka.common.network.ChannelBuilders.clientChannelBuilder(ChannelBuilders.java:73)
        at org.apache.kafka.clients.ClientUtils.createChannelBuilder(ClientUtils.java:105)
        at org.apache.kafka.clients.producer.KafkaProducer.newSender(KafkaProducer.java:442)
        at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:423)
        ... 104 common frames omitted
Caused by: java.util.MissingResourceException: Resource bundle not found sun.security.util.Resources. Register the resource bundle using the option -H:IncludeResourceBundles=sun.security.util.Resources.
        at com.oracle.svm.core.jdk.LocalizationSupport.getCached(LocalizationSupport.java:66)
        at java.util.ResourceBundle.getBundle(ResourceBundle.java:53)
        at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
        at sun.security.util.ResourcesMgr.getBundle(ResourcesMgr.java:54)
        at sun.security.util.ResourcesMgr.getString(ResourcesMgr.java:40)
        at javax.security.auth.login.LoginContext.invoke(LoginContext.java:871)
        at javax.security.auth.login.LoginContext$4.run(LoginContext.java:665)
        at javax.security.auth.login.LoginContext$4.run(LoginContext.java:663)
        at java.security.AccessController.doPrivileged(AccessController.java:146)
        at javax.security.auth.login.LoginContext.invokePriv(LoginContext.java:663)
        at javax.security.auth.login.LoginContext.login(LoginContext.java:579)
        at org.apache.kafka.common.security.authenticator.AbstractLogin.login(AbstractLogin.java:60)
        at org.apache.kafka.common.security.authenticator.LoginManager.<init>(LoginManager.java:62)
        at org.apache.kafka.common.security.authenticator.LoginManager.acquireLoginManager(LoginManager.java:105)
        at org.apache.kafka.common.network.SaslChannelBuilder.configure(SaslChannelBuilder.java:158)
        ... 109 common frames omitted
10:56:31.502 [nioEventLoopGroup-1-3] ERROR i.m.h.s.netty.RoutingInBoundHandler - Unexpected error occurred: Error instantiating bean of type [org.apache.kafka.clients.producer.KafkaProducer]: Failed to construct kafka producer
io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type [org.apache.kafka.clients.producer.KafkaProducer]: Failed to construct kafka producer
        at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1918)
        at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:903)
        at io.micronaut.context.DefaultBeanContext.createBean(DefaultBeanContext.java:830)
        at io.micronaut.context.BeanContext.createBean(BeanContext.java:130)
        at io.micronaut.configuration.kafka.intercept.KafkaClientIntroductionAdvice.lambda$getProducer$13(KafkaClientIntroductionAdvice.java:663)
        at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
        at io.micronaut.configuration.kafka.intercept.KafkaClientIntroductionAdvice.getProducer(KafkaClientIntroductionAdvice.java:585)
        at io.micronaut.configuration.kafka.intercept.KafkaClientIntroductionAdvice.intercept(KafkaClientIntroductionAdvice.java:227)
        at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:69)
        at io.micronaut.retry.intercept.RecoveryInterceptor.intercept(RecoveryInterceptor.java:80)
        at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:69)
        at nl.amis.kafkaclient.TransactionProducer$Intercepted.sendTransaction(Unknown Source)
        at nl.amis.kafkaclient.TransactionProducer.sendTransaction(TransactionProducer.java:11)
        at nl.amis.controller.TransactionController.postMethod(TransactionController.java:25)
        at nl.amis.controller.$TransactionControllerDefinition$Intercepted.$$access0(Unknown Source)
        at nl.amis.controller.$TransactionControllerDefinition$Intercepted$$proxy0.invokeInternal(Unknown Source)
        at io.micronaut.context.AbstractExecutableMethod.invoke(AbstractExecutableMethod.java:146)
        at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:60)
        at io.micronaut.validation.ValidatingInterceptor.intercept(ValidatingInterceptor.java:106)
        at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:69)
        at nl.amis.controller.$TransactionControllerDefinition$Intercepted.postMethod(Unknown Source)
        at nl.amis.controller.$$TransactionControllerDefinition$InterceptedDefinition$$exec1.invokeInternal(Unknown Source)
        at io.micronaut.context.AbstractExecutableMethod.invoke(AbstractExecutableMethod.java:146)
        at io.micronaut.context.DefaultBeanContext$4.invoke(DefaultBeanContext.java:469)
        at io.micronaut.web.router.AbstractRouteMatch.execute(AbstractRouteMatch.java:312)
        at io.micronaut.web.router.RouteMatch.execute(RouteMatch.java:118)
        at io.micronaut.http.server.netty.RoutingInBoundHandler.lambda$buildResultEmitter$9(RoutingInBoundHandler.java:1352)
        at io.reactivex.internal.operators.flowable.FlowableDefer.subscribeActual(FlowableDefer.java:35)
        at io.reactivex.Flowable.subscribe(Flowable.java:14918)
        at io.reactivex.Flowable.subscribe(Flowable.java:14865)
        at io.micronaut.http.server.context.ServerRequestContextFilter.lambda$doFilter$0(ServerRequestContextFilter.java:62)
        at io.reactivex.internal.operators.flowable.FlowableFromPublisher.subscribeActual(FlowableFromPublisher.java:29)
        at io.reactivex.Flowable.subscribe(Flowable.java:14918)
        at io.reactivex.Flowable.subscribe(Flowable.java:14868)
        at io.micronaut.http.server.netty.RoutingInBoundHandler.lambda$buildExecutableRoute$5(RoutingInBoundHandler.java:1056)
        at io.micronaut.web.router.DefaultUriRouteMatch$1.execute(DefaultUriRouteMatch.java:80)
        at io.micronaut.web.router.RouteMatch.execute(RouteMatch.java:118)
        at io.micronaut.http.server.netty.RoutingInBoundHandler$2.doOnComplete(RoutingInBoundHandler.java:987)
        at io.micronaut.core.async.subscriber.CompletionAwareSubscriber.onComplete(CompletionAwareSubscriber.java:71)
        at io.micronaut.http.server.netty.jackson.JsonContentProcessor$1.doOnComplete(JsonContentProcessor.java:140)
        at io.micronaut.core.async.subscriber.CompletionAwareSubscriber.onComplete(CompletionAwareSubscriber.java:71)
        at java.util.Optional.ifPresent(Optional.java:183)
        at io.micronaut.core.async.processor.SingleThreadedBufferingProcessor.doOnComplete(SingleThreadedBufferingProcessor.java:48)
        at io.micronaut.jackson.parser.JacksonProcessor.doOnComplete(JacksonProcessor.java:139)
        at io.micronaut.core.async.subscriber.SingleThreadedBufferingSubscriber.onComplete(SingleThreadedBufferingSubscriber.java:70)
        at io.micronaut.http.server.netty.jackson.JsonContentProcessor.doOnComplete(JsonContentProcessor.java:165)
        at io.micronaut.core.async.subscriber.CompletionAwareSubscriber.onComplete(CompletionAwareSubscriber.java:71)
        at io.micronaut.http.netty.reactive.HandlerPublisher.complete(HandlerPublisher.java:416)
        at io.micronaut.http.netty.reactive.HandlerPublisher.handlerRemoved(HandlerPublisher.java:403)
        at io.netty.channel.AbstractChannelHandlerContext.callHandlerRemoved(AbstractChannelHandlerContext.java:979)
        at io.netty.channel.DefaultChannelPipeline.callHandlerRemoved0(DefaultChannelPipeline.java:637)
        at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:477)
        at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:423)
        at io.micronaut.http.netty.stream.HttpStreamsHandler.removeHandlerIfActive(HttpStreamsHandler.java:388)
        at io.micronaut.http.netty.stream.HttpStreamsHandler.handleReadHttpContent(HttpStreamsHandler.java:253)
        at io.micronaut.http.netty.stream.HttpStreamsHandler.channelRead(HttpStreamsHandler.java:228)
        at io.micronaut.http.netty.stream.HttpStreamsServerHandler.channelRead(HttpStreamsServerHandler.java:121)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
        at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
        at io.netty.handler.codec.http.HttpServerKeepAliveHandler.channelRead(HttpServerKeepAliveHandler.java:64)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.flow.FlowControlHandler.dequeue(FlowControlHandler.java:191)
        at io.netty.handler.flow.FlowControlHandler.read(FlowControlHandler.java:130)
        at io.netty.channel.AbstractChannelHandlerContext.invokeRead(AbstractChannelHandlerContext.java:686)
        at io.netty.channel.AbstractChannelHandlerContext.read(AbstractChannelHandlerContext.java:671)
        at io.netty.channel.ChannelDuplexHandler.read(ChannelDuplexHandler.java:103)
        at io.netty.channel.AbstractChannelHandlerContext.invokeRead(AbstractChannelHandlerContext.java:686)
        at io.netty.channel.AbstractChannelHandlerContext.read(AbstractChannelHandlerContext.java:671)
        at io.netty.channel.ChannelDuplexHandler.read(ChannelDuplexHandler.java:103)
        at io.netty.channel.AbstractChannelHandlerContext.invokeRead(AbstractChannelHandlerContext.java:686)
        at io.netty.channel.AbstractChannelHandlerContext.read(AbstractChannelHandlerContext.java:671)
        at io.netty.channel.ChannelDuplexHandler.read(ChannelDuplexHandler.java:103)
        at io.netty.channel.AbstractChannelHandlerContext.invokeRead(AbstractChannelHandlerContext.java:686)
        at io.netty.channel.AbstractChannelHandlerContext.read(AbstractChannelHandlerContext.java:671)
        at io.micronaut.http.netty.reactive.HandlerPublisher.requestDemand(HandlerPublisher.java:163)
        at io.micronaut.http.netty.stream.HttpStreamsHandler$1.requestDemand(HttpStreamsHandler.java:220)
        at io.micronaut.http.netty.reactive.HandlerPublisher$ChannelSubscription.receivedDemand(HandlerPublisher.java:547)
        at io.micronaut.http.netty.reactive.HandlerPublisher$ChannelSubscription.lambda$request$0(HandlerPublisher.java:474)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:834)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:517)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
Caused by: org.apache.kafka.common.KafkaException: Failed to construct kafka producer
        at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:434)
        at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:313)
        at io.micronaut.configuration.kafka.KafkaProducerFactory.createProducer(KafkaProducerFactory.java:59)
        at io.micronaut.configuration.kafka.$KafkaProducerFactory$CreateProducer0Definition.doBuild(Unknown Source)
        at io.micronaut.context.AbstractParametrizedBeanDefinition.build(AbstractParametrizedBeanDefinition.java:118)
        at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1876)
        ... 99 common frames omitted
Caused by: org.apache.kafka.common.KafkaException: java.util.MissingResourceException: Resource bundle not found sun.security.util.Resources. Register the resource bundle using the option -H:IncludeResourceBundles=sun.security.util.Resources.
        at org.apache.kafka.common.network.SaslChannelBuilder.configure(SaslChannelBuilder.java:172)
        at org.apache.kafka.common.network.ChannelBuilders.create(ChannelBuilders.java:157)
        at org.apache.kafka.common.network.ChannelBuilders.clientChannelBuilder(ChannelBuilders.java:73)
        at org.apache.kafka.clients.ClientUtils.createChannelBuilder(ClientUtils.java:105)
        at org.apache.kafka.clients.producer.KafkaProducer.newSender(KafkaProducer.java:442)
        at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:423)
        ... 104 common frames omitted
Caused by: java.util.MissingResourceException: Resource bundle not found sun.security.util.Resources. Register the resource bundle using the option -H:IncludeResourceBundles=sun.security.util.Resources.
        at com.oracle.svm.core.jdk.LocalizationSupport.getCached(LocalizationSupport.java:66)
        at java.util.ResourceBundle.getBundle(ResourceBundle.java:53)
        at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
        at sun.security.util.ResourcesMgr.getBundle(ResourcesMgr.java:54)
        at sun.security.util.ResourcesMgr.getString(ResourcesMgr.java:40)
        at javax.security.auth.login.LoginContext.invoke(LoginContext.java:871)
        at javax.security.auth.login.LoginContext$4.run(LoginContext.java:665)
        at javax.security.auth.login.LoginContext$4.run(LoginContext.java:663)
        at java.security.AccessController.doPrivileged(AccessController.java:146)
        at javax.security.auth.login.LoginContext.invokePriv(LoginContext.java:663)
        at javax.security.auth.login.LoginContext.login(LoginContext.java:579)
        at org.apache.kafka.common.security.authenticator.AbstractLogin.login(AbstractLogin.java:60)
        at org.apache.kafka.common.security.authenticator.LoginManager.<init>(LoginManager.java:62)
        at org.apache.kafka.common.security.authenticator.LoginManager.acquireLoginManager(LoginManager.java:105)
        at org.apache.kafka.common.network.SaslChannelBuilder.configure(SaslChannelBuilder.java:158)
        ... 109 common frames omitted
```


### Environment Information

- **Operating System**: Windows (Docker Desktop version 2.5.0.0). Error also occurs when running in Google Cloud Run
- **Micronaut Version:** 2.0.1
- **JDK Version:** graalvm-ce-java11-20.2.0

--- a/build.gradle
+++ b/build.gradle
@@ -10,6 +10,11 @@ buildscript {
 
 
 subprojects { Project subproject ->
+
+    if (subproject.path.contains("tests")) {
+        return
+    }
+
     group "io.micronaut.kafka"
 
     apply plugin: "io.micronaut.build.internal.common"

--- a/gradle.properties
+++ b/gradle.properties
@@ -2,10 +2,11 @@ projectVersion=3.3.3-SNAPSHOT
 micronautDocsVersion=2.0.0.RC1
 micronautBuildVersion=1.1.5
 micronautVersion=2.4.2
+micronautGradlePluginVersion=1.5.0
 groovyVersion=3.0.5
 spockVersion=2.0-M3-groovy-3.0
-
 kafkaVersion=2.8.0
+testContainersVersion=1.15.3
 
 title=Micronaut Kafka
 projectDesc=Integration between Micronaut and Kafka Messaging

--- a/kafka-streams/build.gradle
+++ b/kafka-streams/build.gradle
@@ -20,8 +20,8 @@ dependencies {
     testImplementation "org.apache.kafka:kafka-clients:${kafkaVersion}:test"
     testImplementation "org.apache.kafka:kafka_2.12:${kafkaVersion}"
     testImplementation "org.apache.kafka:kafka_2.12:${kafkaVersion}:test"
-    testImplementation "org.testcontainers:kafka:1.15.3" //Breaking change in java container after 1.12 see https://github.com/docker-java/docker-java/pull/1392
-    testImplementation "org.testcontainers:spock:1.15.3"
+    testImplementation "org.testcontainers:kafka:$testContainersVersion"
+    testImplementation "org.testcontainers:spock:$testContainersVersion"
 
     testRuntimeOnly "io.micronaut:micronaut-runtime"
     testRuntimeOnly "io.micronaut:micronaut-http-server-netty"

--- a/kafka/build.gradle
+++ b/kafka/build.gradle
@@ -27,7 +27,7 @@ dependencies {
     testImplementation "org.apache.kafka:kafka_2.12:${kafkaVersion}"
     testImplementation "org.apache.kafka:kafka_2.12:${kafkaVersion}:test"
 
-    testImplementation "org.testcontainers:kafka:1.15.3"
+    testImplementation "org.testcontainers:kafka:$testContainersVersion"
     testImplementation 'io.opentracing.contrib:opentracing-kafka-client:0.1.15'
     testImplementation 'io.opentracing:opentracing-mock:0.33.0'
 

--- a/kafka/src/main/java/io/micronaut/configuration/kafka/graal/ByteBufferUnmapperSubstitute.java
+++ b/kafka/src/main/java/io/micronaut/configuration/kafka/graal/ByteBufferUnmapperSubstitute.java
@@ -1,0 +1,63 @@
+/*
+ * Copyright 2017-2021 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.configuration.kafka.graal;
+
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+import org.apache.kafka.common.utils.ByteBufferUnmapper;
+import org.apache.kafka.common.utils.Java;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.nio.ByteBuffer;
+
+@TargetClass(value = ByteBufferUnmapper.class)
+@SuppressWarnings("MissingJavadocType")
+public final class ByteBufferUnmapperSubstitute {
+
+    @Substitute
+    public static void unmap(String resourceDescription, ByteBuffer buffer) throws IOException {
+        if (!buffer.isDirect()) {
+            throw new IllegalArgumentException("Unmapping only works with direct buffers");
+        }
+
+        try {
+            if (Java.IS_JAVA9_COMPATIBLE) {
+                Class<?> unsafeClass = Class.forName("sun.misc.Unsafe");
+                Method m = unsafeClass.getMethod("cleaner", void.class, ByteBuffer.class);
+                Field f = unsafeClass.getDeclaredField("theUnsafe");
+                f.setAccessible(true);
+                Object theUnsafe = f.get(null);
+                m.invoke(theUnsafe, buffer);
+            } else {
+                Class<?> directBufferClass = Class.forName("java.nio.DirectByteBuffer");
+                Method cleanerMethod = directBufferClass.getMethod("cleaner");
+                cleanerMethod.setAccessible(true);
+                Object cleaner = cleanerMethod.invoke(buffer);
+                if (cleaner != null) {
+                    Class<?> cleanerClass = Class.forName("sun.misc.Cleaner");
+                    Method cleanMethod = cleanerClass.getMethod("clean");
+                    cleanMethod.setAccessible(true);
+                    cleanMethod.invoke(cleaner);
+                }
+            }
+        } catch (Throwable throwable) {
+            throw new IOException("Unable to unmap the mapped buffer: " + resourceDescription, throwable);
+        }
+    }
+
+}

--- a/kafka/src/main/java/io/micronaut/configuration/kafka/graal/SaslClientCallbackHandlerSubstitute.java
+++ b/kafka/src/main/java/io/micronaut/configuration/kafka/graal/SaslClientCallbackHandlerSubstitute.java
@@ -1,0 +1,155 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.configuration.kafka.graal;
+
+import com.oracle.svm.core.annotate.Alias;
+import com.oracle.svm.core.annotate.Inject;
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.config.SaslConfigs;
+import org.apache.kafka.common.security.auth.AuthenticateCallbackHandler;
+import org.apache.kafka.common.security.auth.SaslExtensions;
+import org.apache.kafka.common.security.auth.SaslExtensionsCallback;
+import org.apache.kafka.common.security.authenticator.LoginManager;
+import org.apache.kafka.common.security.authenticator.SaslClientCallbackHandler;
+import org.apache.kafka.common.security.scram.ScramExtensionsCallback;
+import org.apache.kafka.common.security.scram.internals.ScramMechanism;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.security.auth.Subject;
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.NameCallback;
+import javax.security.auth.callback.PasswordCallback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.auth.login.AppConfigurationEntry;
+import javax.security.auth.spi.LoginModule;
+import javax.security.sasl.AuthorizeCallback;
+import javax.security.sasl.RealmCallback;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+// Copy from: https://github.com/oracle/helidon/blob/785ce38ca06b268d16e387fb6498aaaa890695cc/messaging/kafka/src/main/java/io/helidon/messaging/connectors/kafka/SaslClientCallbackHandlerSubstitution.java
+// Workaround for https://github.com/oracle/graal/issues/2745
+@TargetClass(value = SaslClientCallbackHandler.class)
+@SuppressWarnings("MissingJavadocType")
+public final class SaslClientCallbackHandlerSubstitute implements AuthenticateCallbackHandler {
+
+    @Alias
+    private String mechanism;
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private Subject subject;
+
+    @Substitute
+    public SaslClientCallbackHandlerSubstitute() {
+        logger = LoggerFactory.getLogger(LoginManager.class);
+    }
+
+    @Override
+    @Substitute
+    public void configure(Map<String, ?> configs, String saslMechanism, List<AppConfigurationEntry> jaasConfigEntries) {
+        this.mechanism = saslMechanism;
+        this.subject = null;
+
+        int entrySize = jaasConfigEntries.size();
+        if (entrySize == 0) {
+            logger.warn("Missing JAAS config entry, missing or malformed sasl.jaas.config property.");
+            return;
+        } else if (entrySize > 1) {
+            logger.warn("Multiple JAAS config entries, Kafka client's sasl.jaas.config can have only one JAAS config entry.");
+            return;
+        }
+
+        AppConfigurationEntry jaasConfigEntry = jaasConfigEntries.get(0);
+        String jaasLoginModuleName = jaasConfigEntry.getLoginModuleName();
+        subject = new Subject();
+
+        try {
+            Class.forName(jaasLoginModuleName)
+                    .asSubclass(LoginModule.class)
+                    .getDeclaredConstructor()
+                    .newInstance()
+                    .initialize(subject, this, new HashMap<>(), jaasConfigEntry.getOptions());
+        } catch (ReflectiveOperationException e) {
+            throw new KafkaException("Can't instantiate JAAS login module" + jaasLoginModuleName, e);
+        }
+    }
+
+    @Override
+    @Substitute
+    public void handle(Callback[] callbacks) throws UnsupportedCallbackException {
+        // Subject.getSubject doesn't return proper subject in native image
+        // Remove substitution when https://github.com/oracle/graal/issues/2745 is fixed
+        // Subject subject = Subject.getSubject(AccessController.getContext());
+
+        for (Callback callback : callbacks) {
+            if (callback instanceof NameCallback) {
+                NameCallback nc = (NameCallback) callback;
+                if (subject != null && !subject.getPublicCredentials(String.class).isEmpty()) {
+                    nc.setName(subject.getPublicCredentials(String.class).iterator().next());
+                } else {
+                    nc.setName(nc.getDefaultName());
+                }
+            } else if (callback instanceof PasswordCallback) {
+                if (subject != null && !subject.getPrivateCredentials(String.class).isEmpty()) {
+                    char[] password = subject.getPrivateCredentials(String.class).iterator().next().toCharArray();
+                    ((PasswordCallback) callback).setPassword(password);
+                } else {
+                    String errorMessage = "Could not login: the client is being asked for a password, but the Kafka"
+                            + " client code does not currently support obtaining a password from the user.";
+                    throw new UnsupportedCallbackException(callback, errorMessage);
+                }
+            } else if (callback instanceof RealmCallback) {
+                RealmCallback rc = (RealmCallback) callback;
+                rc.setText(rc.getDefaultText());
+            } else if (callback instanceof AuthorizeCallback) {
+                AuthorizeCallback ac = (AuthorizeCallback) callback;
+                String authId = ac.getAuthenticationID();
+                String authzId = ac.getAuthorizationID();
+                ac.setAuthorized(authId.equals(authzId));
+                if (ac.isAuthorized()) {
+                    ac.setAuthorizedID(authzId);
+                }
+            } else if (callback instanceof ScramExtensionsCallback) {
+                if (ScramMechanism.isScram(mechanism) && subject != null && !subject.getPublicCredentials(Map.class).isEmpty()) {
+                    @SuppressWarnings("unchecked")
+                    Map<String, String> extensions =
+                            (Map<String, String>) subject.getPublicCredentials(Map.class).iterator().next();
+                    ((ScramExtensionsCallback) callback).extensions(extensions);
+                }
+            } else if (callback instanceof SaslExtensionsCallback) {
+                if (!SaslConfigs.GSSAPI_MECHANISM.equals(mechanism)
+                        && subject != null && !subject.getPublicCredentials(SaslExtensions.class).isEmpty()) {
+                    SaslExtensions extensions = subject.getPublicCredentials(SaslExtensions.class).iterator().next();
+                    ((SaslExtensionsCallback) callback).extensions(extensions);
+                }
+            } else {
+                throw new UnsupportedCallbackException(callback, "Unrecognized SASL ClientCallback");
+            }
+        }
+    }
+
+    @Override
+    @Substitute
+    public void close() {
+    }
+}

--- a/kafka/src/main/resources/META-INF/native-image/io.micronaut.kafka/micronaut-kafka/reflect-config.json
+++ b/kafka/src/main/resources/META-INF/native-image/io.micronaut.kafka/micronaut-kafka/reflect-config.json
@@ -1,26 +1,135 @@
 [
   {
+    "name":"com.sun.security.sasl.ClientFactoryImpl",
+    "allDeclaredMethods":true,
+    "allDeclaredConstructors":true
+  },
+  {
     "name":"org.apache.kafka.clients.consumer.RangeAssignor",
-    "methods":[{"name":"<init>","parameterTypes":[] }]
+    "allDeclaredMethods":true,
+    "allDeclaredConstructors":true
   },
   {
     "name":"org.apache.kafka.clients.producer.internals.DefaultPartitioner",
-    "methods":[{"name":"<init>","parameterTypes":[] }]
+    "allDeclaredMethods":true,
+    "allDeclaredConstructors":true
   },
   {
     "name":"org.apache.kafka.common.security.authenticator.AbstractLogin$DefaultLoginCallbackHandler",
-    "methods":[{"name":"<init>","parameterTypes":[] }]
+    "allDeclaredMethods":true,
+    "allDeclaredConstructors":true
   },
   {
     "name":"org.apache.kafka.common.security.authenticator.DefaultLogin",
-    "methods":[{"name":"<init>","parameterTypes":[] }]
+    "allDeclaredMethods":true,
+    "allDeclaredConstructors":true
   },
   {
     "name":"org.apache.kafka.common.security.authenticator.SaslClientCallbackHandler",
-    "methods":[{"name":"<init>","parameterTypes":[] }]
+    "allDeclaredMethods":true,
+    "allDeclaredConstructors":true
   },
   {
     "name":"org.apache.kafka.common.security.plain.PlainLoginModule",
+    "allDeclaredMethods":true,
+    "allDeclaredConstructors":true
+  },
+  {
+    "name":"sun.security.pkcs.SignerInfo[]"
+  },
+  {
+    "name":"sun.security.pkcs12.PKCS12KeyStore",
     "methods":[{"name":"<init>","parameterTypes":[] }]
+  },
+  {
+    "name":"sun.security.pkcs12.PKCS12KeyStore$DualFormatPKCS12",
+    "methods":[{"name":"<init>","parameterTypes":[] }]
+  },
+  {
+    "name":"sun.security.provider.JavaKeyStore$JKS",
+    "methods":[{"name":"<init>","parameterTypes":[] }]
+  },
+  {
+    "name":"sun.security.provider.NativePRNG",
+    "methods":[{"name":"<init>","parameterTypes":[] }]
+  },
+  {
+    "name":"sun.security.provider.SHA",
+    "methods":[{"name":"<init>","parameterTypes":[] }]
+  },
+  {
+    "name":"sun.security.provider.X509Factory",
+    "methods":[{"name":"<init>","parameterTypes":[] }]
+  },
+  {
+    "name":"sun.security.rsa.RSAKeyFactory$Legacy",
+    "methods":[{"name":"<init>","parameterTypes":[] }]
+  },
+  {
+    "name":"sun.security.ssl.SSLContextImpl$TLSContext",
+    "methods":[{"name":"<init>","parameterTypes":[] }]
+  },
+  {
+    "name":"sun.security.ssl.TrustManagerFactoryImpl$PKIXFactory",
+    "methods":[{"name":"<init>","parameterTypes":[] }]
+  },
+  {
+    "name":"sun.security.util.ObjectIdentifier"
+  },
+  {
+    "name":"sun.security.x509.AuthorityInfoAccessExtension",
+    "methods":[{"name":"<init>","parameterTypes":["java.lang.Boolean","java.lang.Object"] }]
+  },
+  {
+    "name":"sun.security.x509.AuthorityKeyIdentifierExtension",
+    "methods":[{"name":"<init>","parameterTypes":["java.lang.Boolean","java.lang.Object"] }]
+  },
+  {
+    "name":"sun.security.x509.BasicConstraintsExtension",
+    "methods":[{"name":"<init>","parameterTypes":["java.lang.Boolean","java.lang.Object"] }]
+  },
+  {
+    "name":"sun.security.x509.CRLDistributionPointsExtension",
+    "methods":[{"name":"<init>","parameterTypes":["java.lang.Boolean","java.lang.Object"] }]
+  },
+  {
+    "name":"sun.security.x509.CertificateExtensions"
+  },
+  {
+    "name":"sun.security.x509.CertificatePoliciesExtension",
+    "methods":[{"name":"<init>","parameterTypes":["java.lang.Boolean","java.lang.Object"] }]
+  },
+  {
+    "name":"sun.security.x509.ExtendedKeyUsageExtension",
+    "methods":[{"name":"<init>","parameterTypes":["java.lang.Boolean","java.lang.Object"] }]
+  },
+  {
+    "name":"sun.security.x509.IssuerAlternativeNameExtension",
+    "methods":[{"name":"<init>","parameterTypes":["java.lang.Boolean","java.lang.Object"] }]
+  },
+  {
+    "name":"sun.security.x509.KeyUsageExtension",
+    "methods":[{"name":"<init>","parameterTypes":["java.lang.Boolean","java.lang.Object"] }]
+  },
+  {
+    "name":"sun.security.x509.NetscapeCertTypeExtension",
+    "methods":[{"name":"<init>","parameterTypes":["java.lang.Boolean","java.lang.Object"] }]
+  },
+  {
+    "name":"sun.security.x509.PrivateKeyUsageExtension",
+    "methods":[{"name":"<init>","parameterTypes":["java.lang.Boolean","java.lang.Object"] }]
+  },
+  {
+    "name":"sun.security.x509.SubjectAlternativeNameExtension",
+    "methods":[{"name":"<init>","parameterTypes":["java.lang.Boolean","java.lang.Object"] }]
+  },
+  {
+    "name":"sun.security.x509.SubjectKeyIdentifierExtension",
+    "methods":[{"name":"<init>","parameterTypes":["java.lang.Boolean","java.lang.Object"] }]
+  },
+  {
+    "name":"org.apache.commons.compress.archivers.ArchiveOutputStream",
+    "allDeclaredMethods":true,
+    "allDeclaredConstructors":true
   }
 ]

--- a/settings.gradle
+++ b/settings.gradle
@@ -1,4 +1,12 @@
+pluginManagement {
+    plugins {
+        id 'io.micronaut.application' version getProperty("micronautGradlePluginVersion")
+        id 'io.micronaut.library' version getProperty("micronautGradlePluginVersion")
+    }
+}
+
 rootProject.name = 'kafka'
 
 include 'kafka'
 include 'kafka-streams'
+include 'tests:tasks-sasl-plaintext'

--- a/tests/tasks-sasl-plaintext/build.gradle
+++ b/tests/tasks-sasl-plaintext/build.gradle
@@ -1,0 +1,27 @@
+plugins {
+    id "java"
+    id "groovy"
+    id "io.micronaut.application"
+}
+
+application {
+    mainClass = "example.Application"
+}
+
+micronaut {
+    version project.ext.micronautVersion
+    runtime "netty"
+    testRuntime "spock"
+}
+
+repositories {
+    mavenCentral()
+}
+
+dependencies {
+    implementation project(":kafka")
+    testImplementation "io.micronaut:micronaut-http-client"
+    testImplementation platform("org.testcontainers:testcontainers-bom:$testContainersVersion")
+    testImplementation "org.testcontainers:kafka"
+    testImplementation "org.testcontainers:spock"
+}

--- a/tests/tasks-sasl-plaintext/src/main/java/example/Application.java
+++ b/tests/tasks-sasl-plaintext/src/main/java/example/Application.java
@@ -1,0 +1,11 @@
+
+package example;
+
+import io.micronaut.runtime.Micronaut;
+
+public class Application {
+
+    public static void main(String[] args) {
+        Micronaut.run(Application.class);
+    }
+}

--- a/tests/tasks-sasl-plaintext/src/main/java/example/Task.java
+++ b/tests/tasks-sasl-plaintext/src/main/java/example/Task.java
@@ -1,0 +1,17 @@
+package example;
+
+import io.micronaut.core.annotation.Introspected;
+
+@Introspected
+public class Task {
+
+    private final int id;
+
+    public Task(int id) {
+        this.id = id;
+    }
+
+    public int getId() {
+        return id;
+    }
+}

--- a/tests/tasks-sasl-plaintext/src/main/java/example/TaskConstants.java
+++ b/tests/tasks-sasl-plaintext/src/main/java/example/TaskConstants.java
@@ -1,0 +1,7 @@
+package example;
+
+public interface TaskConstants {
+
+    String TOPIC = "tasks";
+
+}

--- a/tests/tasks-sasl-plaintext/src/main/java/example/TasksController.java
+++ b/tests/tasks-sasl-plaintext/src/main/java/example/TasksController.java
@@ -1,0 +1,14 @@
+package example;
+
+import io.micronaut.http.annotation.Controller;
+import io.micronaut.http.annotation.Get;
+
+@Controller("/tasks")
+public class TasksController {
+
+    @Get("/processed-count")
+    public int get() {
+        return TasksListener.TASKS_PROCESSED.get();
+    }
+
+}

--- a/tests/tasks-sasl-plaintext/src/main/java/example/TasksListener.java
+++ b/tests/tasks-sasl-plaintext/src/main/java/example/TasksListener.java
@@ -1,0 +1,25 @@
+package example;
+
+import io.micronaut.configuration.kafka.annotation.KafkaListener;
+import io.micronaut.configuration.kafka.annotation.OffsetReset;
+import io.micronaut.configuration.kafka.annotation.Topic;
+import io.micronaut.messaging.annotation.MessageBody;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+@KafkaListener(offsetReset = OffsetReset.EARLIEST)
+public class TasksListener {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TasksListener.class);
+
+    public static final AtomicInteger TASKS_PROCESSED = new AtomicInteger();
+
+    @Topic(TaskConstants.TOPIC)
+    public void receive(@MessageBody Task task) {
+        LOGGER.info("Received task with id: " + task.getId());
+        TASKS_PROCESSED.incrementAndGet();
+    }
+
+}

--- a/tests/tasks-sasl-plaintext/src/main/java/example/TasksProducer.java
+++ b/tests/tasks-sasl-plaintext/src/main/java/example/TasksProducer.java
@@ -1,0 +1,13 @@
+package example;
+
+import io.micronaut.configuration.kafka.annotation.KafkaClient;
+import io.micronaut.configuration.kafka.annotation.Topic;
+import io.micronaut.messaging.annotation.MessageBody;
+
+@KafkaClient
+public interface TasksProducer {
+
+    @Topic(TaskConstants.TOPIC)
+    void send(@MessageBody Task body);
+
+}

--- a/tests/tasks-sasl-plaintext/src/main/java/example/TasksPublisher.java
+++ b/tests/tasks-sasl-plaintext/src/main/java/example/TasksPublisher.java
@@ -1,0 +1,32 @@
+package example;
+
+import io.micronaut.scheduling.annotation.Scheduled;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Singleton;
+import java.util.concurrent.atomic.AtomicInteger;
+
+@Singleton
+public class TasksPublisher {
+
+    private final static Logger LOGGER = LoggerFactory.getLogger(TasksPublisher.class);
+    private final AtomicInteger tasksIds = new AtomicInteger();
+    private final TasksProducer tasksProducer;
+
+    public TasksPublisher(TasksProducer tasksProducer) {
+        this.tasksProducer = tasksProducer;
+    }
+
+    @Scheduled(initialDelay = "2s", fixedRate = "1s")
+    public void sendMessages() {
+        try {
+            int id = tasksIds.incrementAndGet();
+            LOGGER.info("Publishing a task with id: " + id);
+            tasksProducer.send(new Task(id));
+        } catch (Exception e) {
+            LOGGER.error("Failed to publish a task", e);
+        }
+    }
+
+}

--- a/tests/tasks-sasl-plaintext/src/main/resources/application.yml
+++ b/tests/tasks-sasl-plaintext/src/main/resources/application.yml
@@ -1,0 +1,8 @@
+kafka:
+  security:
+    protocol: SASL_PLAINTEXT
+  sasl:
+    mechanism: PLAIN
+    jaas:
+      config: org.apache.kafka.common.security.plain.PlainLoginModule required username="admin" password="admin-secret";
+

--- a/tests/tasks-sasl-plaintext/src/main/resources/logback.xml
+++ b/tests/tasks-sasl-plaintext/src/main/resources/logback.xml
@@ -1,0 +1,17 @@
+<configuration>
+
+    <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
+        <withJansi>true</withJansi>
+        <!-- encoders are assigned the type
+             ch.qos.logback.classic.encoder.PatternLayoutEncoder by default -->
+        <encoder>
+            <pattern>%cyan(%d{HH:mm:ss.SSS}) %gray([%thread]) %highlight(%-5level) %magenta(%logger{36}) - %msg%n</pattern>
+        </encoder>
+    </appender>
+
+    <root level="info">
+        <appender-ref ref="STDOUT" />
+    </root>
+
+
+</configuration>

--- a/tests/tasks-sasl-plaintext/src/test/groovy/example/TasksSpec.groovy
+++ b/tests/tasks-sasl-plaintext/src/test/groovy/example/TasksSpec.groovy
@@ -1,0 +1,49 @@
+package example
+
+
+import io.micronaut.http.client.HttpClient
+import io.micronaut.http.client.annotation.Client
+import io.micronaut.test.extensions.spock.annotation.MicronautTest
+import io.micronaut.test.support.TestPropertyProvider
+import org.testcontainers.containers.KafkaContainer
+import org.testcontainers.utility.MountableFile
+import spock.lang.AutoCleanup
+import spock.lang.Shared
+import spock.lang.Specification
+import spock.util.concurrent.PollingConditions
+
+import javax.inject.Inject
+
+@MicronautTest
+class TasksSpec extends Specification implements TestPropertyProvider {
+
+    @Shared
+    @AutoCleanup
+    KafkaContainer kafkaContainer = new KafkaContainer()
+            .withEnv("KAFKA_LISTENER_SECURITY_PROTOCOL_MAP", "BROKER:PLAINTEXT,PLAINTEXT:SASL_PLAINTEXT")
+            .withEnv("KAFKA_SASL_ENABLED_MECHANISMS", "PLAIN")
+            .withEnv("KAFKA_SASL_MECHANISM_INTER_BROKER_PROTOCOL", "PLAIN")
+            .withEnv("KAFKA_OPTS", "-Djava.security.auth.login.config=/etc/kafka/kafka_server_jaas.conf")
+            .withCopyFileToContainer(MountableFile.forClasspathResource("/example/kafka_server_jaas.conf"), "/etc/kafka/kafka_server_jaas.conf")
+
+    @Inject
+    @Client("/")
+    HttpClient client
+
+    PollingConditions pollingConditions = new PollingConditions(initialDelay: 2, timeout: 100)
+
+    def 'should process tasks'() {
+        expect:
+            pollingConditions.eventually {
+                client.toBlocking().retrieve("/tasks/processed-count", Integer) > 3
+            }
+    }
+
+    @Override
+    Map<String, String> getProperties() {
+        kafkaContainer.start()
+        return [
+                'kafka.bootstrap.servers': kafkaContainer.bootstrapServers
+        ]
+    }
+}

--- a/tests/tasks-sasl-plaintext/src/test/resources/example/kafka_server_jaas.conf
+++ b/tests/tasks-sasl-plaintext/src/test/resources/example/kafka_server_jaas.conf
@@ -1,0 +1,7 @@
+KafkaServer {
+  org.apache.kafka.common.security.plain.PlainLoginModule required
+  username="admin"
+  password="admin-secret"
+  user_admin="admin-secret";
+};
+Client {};

